src/overlay/TCPPeer.h

```diff
@@ -102,6 +119,8 @@ class TCPPeer : public Peer
 
     virtual ~TCPPeer();
 
+    // Thread-safe peer drop (can be initiated from either main or overlay
+    // thread)
     virtual void drop(std::string const& reason, DropDirection dropDirection,
```

Comment:
Drop can be initiated from either thread, because there are different reasons for dropping a connection. E.g. we may encounter a super low-level ASIO error from background, or main thread deciding that peer doesn't follow the protocol at that (higher) layer of abstraction. That being said, in the updated version there is no need to synchronize peer lists, and the responsibilities are separated such that peer list is always handled on main, while socket shutdown is handled in the background.